man/alist.man
.\" 3.2
.\"  /*******************************************************************\
.\"   * Copyright (C) 1995 Lars J Aas <larsa@colargol.stud.idb.hist.no> *
.\"   * based on documentation by C J Needham <col@imdb.com> 1995,      *
.\"   * permission is granted by the authors to freely distribute       *
.\"   *                        providing no fee of any kind is charged. *
.\"  \*******************************************************************/
.\"
.TH ALIST 1 "RELEASEDATE" " " "The Internet Movie Database vVERSIONNUMBER"
.SH NAME
alist \- search the attribute fields of all the databases in the movie database.
.SH SYNOPSYS
.B
alist [options]
.SH DESCRIPTION
This program searches the attribute fields of all the databases. Examples
of the kind of queries you can construct include:
.PP
.TP
* search for details of all best actor academy award winners
.TP
* search for all movies nominated for best picture
.TP
* search for all movies narrated by a particular person
.TP
* search for all plays by a particular writer that have been filmed
.SH DATABASE OPTIONS
The program can be run in title-mode and in name-mode.
For title-mode just use:
.TP
.B \-title
.PP
In name-mode,
.B one
of the following options can be used:
.TP
.B \-acr
actors database
.TP
.B \-acs
actresses database
.TP
.B \-cast
both cast databases
.TP
.B \-dir
directors database
.TP
.B \-write
writers database
.TP
.B \-comp
composers database
.TP
.B \-cine
cinematographers database
.TP
.B \-edit
editors database
.TP
.B \-prodes
production designers database
.TP
.B \-costdes
costume designers database
.TP
.B \-prdcr
producers database
.TP
.B \-misc
miscellaneous database
.SH SEARCH OPTIONS
.TP
.B \-aa
academy award winners
.TP
.B \-aan
academy award nominees
.TP
.B \-gg
golden globe winners
.TP
.B \-ggn
golden globe nominees
.TP
.B \-attr <pattern>
values matching
.B <pattern>
.PP
These options can be used to limit the search in various ways:
.TP
.B \-m
Restrict search to movies only.
.TP
.B \-yrfrom <yr>
Restrict search to titles released during or after
.B <yr>
.TP
.B \-yrto <yr>
Restrict search to titles released before
.B <yr>
.TP
.B \-yreq <yr>
Restrict search to titles released in
.B <yr>
.SH MISCELLANEOUS OPTIONS
.TP
.B \-s
Substring: allow partial matches with
.BR \-attr .
.TP
.B \-i
Become case sensitive in
.B \-attr
searches.
.TP
.B \-yr
Adds year of release to titles and group results by year.
.SH EXAMPLES
.TP
Search directors database for academy award winners/nominees and sort by year.
.B
alist \-dir \-aa \-aan \-yr
.TP
Search for best picture academy award winners/nominees and sort by year.
.B
alist \-title \-aa \-aan \-yr
.TP
Lists all best director oscar winners of the 1940s by year.
.B
alist \-aa \-dir \-yr \-yrfrom 1940 \-yrto 1949


docs/ChangeLog
-------Version 3.4 Released---------------------

Sun Mar 29 22:40:37 BST 1998

 * biographies list offset fields increased to 4 bytes following overflow
 * various constants updated

-------Version 3.3c Released---------------------

Sat Apr 26 06:51:19 BST 1997

 * added support for crew completion list
 * added support for distributors list
 * adapted mkdbs for more robust processing of filmography lists
 * added support for new business list tag

-------Version 3.3b Released---------------------

Sat Mar 15 08:09:50 GMT 1997

 * added support for ISO-Latin-1

-------Version 3.3 Released---------------------

Wed Oct 16 22:28:54 BST 1996

 * added support for AKA attributes
 * added support for writers ordering
 * moved misc section to after case in title output
 * added support for business list
 * added support for laserdisc list
 * fixed tagline bug
 * diplays URL of web page for name/title

-------Version 3.2h Released---------------------

Tue Oct  8 14:24:21 BST 1996

 * added support for special effects companies list
 * updated various system constants
 * added new movie link types 'edited into' & 'edited from'
 * added extra checks to avoid title/attribute key file corruption in mkdbs.c

-------Version 3.2f Released---------------------

Mon Jun 24 22:04:46 GMT 1996

 * added support for language list

-------Version 3.2e Released---------------------

Sat Mar 16 08:09:39 GMT 1996

 * added support for alternate versions list
 * added bio keys IT & AT

-------Version 3.2d Released---------------------

Tue Aug 15 23:23:47 GMT 1995

 * switched TitleID to unsigned long to handle >65K titles
 * fixed bug in display of literature and genre data
 * added bio key WN

-------Version 3.2c Released---------------------

Sun May 14 23:00:14 GMT 1995

 * raised MAXFILMOGRAPHIES constant to handle larger no. of entries

-------Version 3.2 Released---------------------

Sun Jan 29 22:40:08 GMT 1995

 * added full support for title searching without year/TV/V info

Tue Jan  3 23:00:45 GMT 1995

 * replaced 'biodefs' file
 * fixed bug when looking for common titles with people using -name

Mon Jan  2 21:52:08 GMT 1995

 * log file support
 * better separation of search/display code

Wed Dec 14 21:43:57 GMT 1994

 * fixed bug in 'title' for miscellaneous section
 * added support for the movie links list

Tue Dec 13 22:06:44 GMT 1994
 
 * noted when cast in credits output
 * fixed name ordering problem for names with (I)/(II) marker
 * display Actor/Actress in list -name where appropriate

Wed Dec  7 21:19:19 GMT 1994

 * added support for tag lines & complete cast list
 * put spaces around "." in cast results
 * standardized number of dashes printed out in 'title' / 'list'

Mon Dec  5 21:19:42 GMT 1994

 * multi-country productions displayed on same line

Wed Nov 30 18:14:16 GMT 1994

 * switched to new IMDB license
 * updated 'lfetch' to get new lists
 * fixed "No plot summaries message"
 * simplified MRR processing in mkdb
 * added support for locations & release date lists
 * switched 'Reviews' marker for clarity
 * added support for technical & literature lists

Mon Oct 31 21:42:03 GMT 1994
 
-------Version 3.1 Released---------------------

Mon Oct 17 19:04:26 GMT 1994
 
 * 3.1 code frozen
 * minor bug fixes with lindex

Sun Oct  9 18:37:14 GMT 1994

 * added support for countries list

Tue Oct  4 21:27:53 GMT 1994

 * added support for cast ordering info

Sun Oct  2 23:02:59 GMT 1994

 * added support for color info and sound mix lists
 * added constraint options to 'list'
 * standardized list options

Wed Sep 28 22:22:56 GMT 1994

 * re-ordered title info printing plus separated genres list
 * print genres on one line

Mon Sep 26 21:53:47 GMT 1994

 * added support for production companies list

Sun Sep 25 22:02:07 GMT 1994

 * added constraint options to 'lindex'

Tue Sep  6 20:09:52 GMT 1994

 * added support for certificates & genres list
 * fixed mkballot bug with -ar option

Tue Aug 23 20:08:04 GMT 1994

 * added support for miscellaneous filmographies list

Tue May 24 19:13:47 GMT 1994

 * moviedb-3.0 completed

Thu May 19 22:35:12 GMT 1994

 * switched to merged character names scheme

Tue May 17 21:09:45 GMT 1994

 * new 'mkdb' eliminating need for large array to hold name keys
 * removed duplicate aka titles in 'title -s -aka' queries

Wed May 11 22:11:59 GMT 1994

 * made upperCase a macro for increased speed
 * switched to Phil's new Makefiles

Wed May  4 17:46:08 GMT 1994

 * mapped case of titles/names to match key files
 * added support for running times list
 * lguide error messages

Tue May  3 21:38:30 GMT 1994
 
 * added support for Phil's imoviedb package
 * expanded compression support

Thu Apr 28 19:51:52 GMT 1994

 * finished rewrite to use new database representation

-------Version 2.9 Released---------------------

Wed Mar 16 21:39:40 GMT 1994

 * added support for producers list
 * added support for soundtracks list
 * added support for generic aka lists

Tue Mar 15 23:47:46 GMT 1994

 * replaced -crazy and -triv options with -trivia (also includes goofs & quotes)
 * added support for quotes list
 * added support for goofs list
 * handled -full error messages correctly
 * correct conditions for printing of missing plot summaries message
 * fixed lindex to search all lists when no list options given
 * -mrr no longer implies -m

Mon Mar 14 21:48:21 GMT 1994

 * new version of check-for-program
 * passed $MAKE variable to make subprocesses
 * lfetch modified to match full filenames
 * added 'keepchars' option for CLEANLISTS

-------Version 2.8 Released---------------------

Wed Dec  1 21:58:21 GMT 1993

 * man pages added to package
 * LaTeX documentation added to package

Sat Nov 27 20:53:40 GMT 1993

 * 2.8 code finished!

Thu Nov 25 18:17:00 GMT 1993

 * ported lseen to C

Tue Nov 23 19:16:48 GMT 1993

 * added gendb -votes option

Mon Nov 22 23:54:58 GMT 1993

 * ported lformat to C
 * removed yrf.c file

Wed Nov 17 22:36:44 GMT 1993


src/alist.c
/*============================================================================
 *
 *  Program: alist.c 
 *
 *  Version: 3.7
 *
 *  Author:  C J Needham <col@imdb.com>
 *
 *  Copyright (c) 1996-1999 The Internet Movie Database Inc
 *
 *  Permission is granted by the copyright holder to distribute this program
 *  is source form only, providing this notice remains intact, and no fee
 *  of any kind is charged. This software is provided as is and there are no 
 *  warranties, expressed or implied.
 * 
 *  Options:
 *
 *        -acr  search actors database
 *        -acs  search actress database
 *        -dir  search directors database
 *      -write  search writers database
 *       -comp  search composers database
 *       -cine  search cinematographers database
 *       -edit  search editors database
 *     -prodes  search production designers database
 *    -costdes  search costume designers database
 *      -prdcr  search producers database
 *       -misc  search miscellaneous database
 *
 *      -title  search titles database
 *
 *   -attr <a>  search for entries with attribute equal to string <a>
 *         -aa  search for academy award winners
 *        -aan  search for academy award nominess
 *         -gg  search for Golden Globe winners
 *        -ggn  search for Golden Globe nominees
 *
 *         -yr  add year of release info & sort chronologically
 * -yrfrom <y>  limit search to movies released from year <y> onwards
 *   -yrto <y>  limit search to movies released before year <y>
 *   -yreq <y>  limit search to movies released in year <y>
 *
 *          -m  movies only, ignore TV-series
 *          -i  case sensistive search
 *          -s  enable substring matching for -attr value only
 *
 *============================================================================
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "log.h"
#include "moviedb.h"
#include "dbutils.h"
#include "years.h"

#define ALIST_USAGE1 "usage: alist [-acr -acs -dir -write -comp -cine -edit -prodes -costdes -prdcr" 
#define ALIST_USAGE2 "              -misc] [-title] [-yr -yrfrom <year> -yrto <year> -yreq <year>]"
#define ALIST_USAGE3 "             [-m -i -s] -attr <attr>|{-aa -aan}|{-gg -ggn}"

int alistYearSort ( struct alistResult *r1, struct alistResult *r2 )
{
  if ( r1 -> year != r2 -> year )
    return ( r1 -> year - r2 -> year ) ;
  else
    return ( caseCompare ( r1 -> title, r2 -> title ) ) ;
}


void displayAlistSorted ( struct alistResult *results, int count ) 
{
  int i, len, currentYear = 0 ;

  qsort ( (void*) results, (size_t) count, sizeof ( struct alistResult ), (int (*) (const void*, const void*)) alistYearSort ) ;
  
  for ( i = 0 ; i < count ; i++ )
  {
    if ( currentYear != results [ i ] . year )
    {
      (void) printf ( "\n" ) ;
      currentYear = results [ i ] . year ;
    }
    (void) printf ( "%s", results [ i ] . name ) ;
    len = strlen ( results [ i ] . name ) ;
    if ( len >= 16 )
      (void) printf ( "\t" ) ;
    else if ( len >= 8 )
      (void) printf ( "\t\t" ) ;
    else
      (void) printf ( "\t\t\t" ) ;

    displayTitleYear ( results [ i ] . title, results [ i ] . year ) ;
    if ( results [ i ] . attr != NULL )
      (void) printf ( " %s\n", results [ i ] . attr ) ;
    else
      (void) printf ( "\n" ) ;
  }
}


int alistAttrTest ( struct attrSearchOptRec attrFlags, char *attr, int substring, int casesen, char *listAttr ) 
{
  if ( listAttr == NULL )
    return ( FALSE ) ;

  if ( attrFlags . string )
    if ( substring )
    {
      if ( casesen )
      {
        if ( strstr ( listAttr, attr ) != NULL )
          return ( TRUE ) ;
      }
      else
        if ( caseStrStr ( listAttr, attr ) != NULL )
          return ( TRUE ) ;
    }
    else
    {
      if ( casesen )
      {
        if ( strcmp ( listAttr, attr ) == 0 )
          return ( TRUE ) ;
      }
      else
        if ( caseCompare ( listAttr, attr ) == 0 )
          return ( TRUE ) ;
    }      

  if ( attrFlags . aa )
    if ( strstr ( listAttr, "AA)" ) != NULL )
      return ( TRUE ) ;

  if ( attrFlags . aan )
    if ( strstr ( listAttr, "AAN)" ) != NULL )
      return ( TRUE ) ;

  if ( attrFlags . gg )
    if ( strstr ( listAttr, "GG)" ) != NULL )
      return ( TRUE ) ;

  if ( attrFlags . ggn )
    if ( strstr ( listAttr, "GGN)" ) != NULL )
      return ( TRUE ) ;

  return ( FALSE ) ;
}


int titleKeyAlistCompare ( struct titleDbRec *search, struct titleDbRec *t1 )
{
  return ( search -> titleKey - t1 -> titleKey ) ;
}


int alistYearTest ( int yearFlag, int yrf, int yrt, TitleID titleKey, TitleID nyears, struct titleDbRec *years, int *yearValue ) 
{
  struct titleDbRec *trec, searchRec ;

  if ( yearFlag )
  {
    searchRec . titleKey = titleKey ;
    if ( ( trec = bsearch ( (void*) &searchRec, (void*) years, (size_t) nyears, sizeof ( struct titleDbRec ), (int (*) (const void*, const void*)) titleKeyAlistCompare ) ) != NULL )
      *yearValue = trec -> year ;
    else
      return ( FALSE ) ;
    if ( *yearValue < yrf )
      return ( FALSE ) ;
    if ( *yearValue > yrt )
      return ( FALSE ) ;
    return ( TRUE ) ;
  }
  *yearValue = 0 ;
  return ( TRUE ) ;
}


int alistMoviesOnlyTest ( int mvsonly, char *title )
{
  if ( mvsonly && *title == '"' )
    return ( FALSE ) ;
  else
    return ( TRUE ) ;
}


void alistNames ( int listId, char *attr, int substring, int casesen, struct attrSearchOptRec attrFlags, int mvsonly, int yr, int yrf, int yrt, TitleID nyears, struct titleDbRec *years )
{
  FILE *dbFp, *nameKeyFp, *nameIndexFp, *titleKeyFp, *titleIndexFp, *attrKeyFp, *attrIndexFp ;
  struct formatRec listData [ MAXFILMOGRAPHIES ] ;
  struct alistResult results [ MAXALISTRESULTS ] ;
  int i, count, len, noWithAttr, noWithoutAttr, rcount = 0 ;
  char fn [ MAXPATHLEN ] ;
  char *name ;
  NameID nameKey ;
  TitleID titleKey ;
  AttributeID attrKey ;
  int yearValue, keepFlag ;

  (void) constructFilename ( fn, filmographyDefs [ listId ] .stem, DBSEXT ) ;
  dbFp = openFile ( fn ) ;
  nameKeyFp = openFile ( NAMEKEY ) ;
  nameIndexFp = openFile ( NAMEIDX ) ;
  titleKeyFp = openFile ( TITLEKEY ) ;
  titleIndexFp = openFile ( TITLEIDX ) ;
  attrKeyFp = openFile ( ATTRKEY ) ;
  attrIndexFp = openFile ( ATTRIDX ) ;

  while ( TRUE )
  {
    count = 0 ;
    nameKey = getName ( dbFp ) ;
    name = mapNameKeyToText ( nameKey, nameKeyFp, nameIndexFp ) ;
    if ( feof ( dbFp ) )
      break ;
    getFilmographyCounts ( dbFp, &noWithAttr, &noWithoutAttr ) ;
    for ( i = 0 ; i < noWithAttr ; i++ )
    {
      titleKey = getTitle ( dbFp ) ;
      attrKey = getAttr ( dbFp ) ;
      if ( listId < NO_OF_CAST_LISTS )
      {
        len = getByte ( dbFp ) ;
        if ( len > 0 )
          (void) fseek ( dbFp, len, SEEK_CUR ) ;
        (void) getPosition ( dbFp ) ;
      }
	else if ( listId == WRITER_LIST_ID )
      {
	(void) getPosition ( dbFp );
	(void) getPosition ( dbFp );
	(void) getPosition ( dbFp );
      }
      listData [ count ] . titleKey = titleKey ;
      listData [ count ] . attrKey = attrKey ;
      listData [ count ] . title = mapTitleKeyToText ( titleKey, titleKeyFp, titleIndexFp ) ;
      listData [ count ] . attr = mapAttrKeyToText ( attrKey, attrKeyFp, attrIndexFp ) ;
      keepFlag = alistMoviesOnlyTest ( mvsonly, listData [ count ] . title ) ;
      keepFlag = keepFlag && alistYearTest ( yr, yrf, yrt, titleKey, nyears, years, &yearValue ) ;
      keepFlag = keepFlag && alistAttrTest ( attrFlags, attr, substring, casesen, listData [ count ] . attr ) ;
      listData [ count ] . year = yearValue ;
      if ( keepFlag )
      {
        if ( yr )
        {
          results [ rcount ] . name = duplicateString ( name ) ;
          results [ rcount ] . title = listData [ count ] . title ;
          results [ rcount ] . attr = listData [ count ] . attr ;
          results [ rcount ] . year = yearValue ;
          rcount++ ;
        }
        count++ ;
      }
      else
      {
        free ( (void*) listData [ count ] . title ) ;
        free ( (void*) listData [ count ] . attr ) ;
      }
    }
    for ( i = 0 ; i < noWithoutAttr ; i++ )
    {
      (void) getTitle ( dbFp ) ;
      if ( listId < NO_OF_CAST_LISTS )
      {
        len = getByte ( dbFp ) ;
        if ( len > 0 )
          (void) fseek ( dbFp, len, SEEK_CUR ) ;
        (void) getPosition ( dbFp ) ;
      }
	else if ( listId == WRITER_LIST_ID )
      {
	(void) getPosition ( dbFp );
	(void) getPosition ( dbFp );
	(void) getPosition ( dbFp );
      }
    }
    if ( ! yr )
    {
      listData [ 0 ] . name = name ;
      displayFormatPlain ( listData, count ) ;
      freeFormatData ( listData, count ) ;
    }
  }
  if ( yr )
  {
    displayAlistSorted ( results, rcount ) ;
    for ( i = 0 ; i < rcount ; i++ )
    {
      free ( (void*) results [ i ] . name ) ;
      free ( (void*) results [ i ] . title ) ;
      free ( (void*) results [ i ] . attr ) ;
    }
  }
  (void) fclose ( dbFp ) ;
  (void) fclose ( nameKeyFp ) ;
  (void) fclose ( nameIndexFp ) ;
  (void) fclose ( titleKeyFp ) ;
  (void) fclose ( titleIndexFp ) ;
  (void) fclose ( attrKeyFp ) ;
  (void) fclose ( attrIndexFp ) ;
}


int alistTitleYearSort ( struct alistTitleRes *r1, struct alistTitleRes *r2 )
{
  if ( r1 -> year != r2 -> year )
    return ( r1 -> year - r2 -> year ) ;
  else
    return ( caseCompare ( r1 -> title, r2 -> title ) ) ;
}


void displayAlistTitleSorted ( struct alistTitleRes results [], int count ) 
{
  int i, currentYear = 0 ;

  qsort ( (void*) results, (size_t) count, sizeof ( struct alistTitleRes ), (int (*) (const void*, const void*)) alistTitleYearSort ) ;
  
  for ( i = 0 ; i < count ; i++ )
  {
    if ( currentYear != results [ i ] . year )
    {
      (void) printf ( "\n" ) ;
      currentYear = results [ i ] . year ;
    }
    displayTitleYear ( results [ i ] . title, results [ i ] . year ) ;
    if ( results [ i ] . attr != NULL )
      (void) printf ( " %s\n", results [ i ] . attr ) ;
    else
      (void) printf ( "\n" ) ;
  }
}


void alistTitles ( char *attr, int substring, int casesen, struct attrSearchOptRec attrFlags, int mvsonly, int yr, int yrf, int yrt )
{
  FILE *dbFp, *titleKeyFp, *titleIndexFp, *attrKeyFp, *attrIndexFp ;
  char *title, *dbAttr ;
  TitleID titleKey ;
  AttributeID attrKey ;
  int  year, keepFlag ;
  struct alistTitleRes results [ MAXALISTTITLES ] ;
  int count = 0, i;

  dbFp = openFile ( MOVIEDB ) ;
  titleKeyFp = openFile ( TITLEKEY ) ;
  titleIndexFp = openFile ( TITLEIDX ) ;
  attrKeyFp = openFile ( ATTRKEY ) ;
  attrIndexFp = openFile ( ATTRIDX ) ;

  while ( TRUE )
  {
    titleKey = getTitle ( dbFp ) ;
    if ( feof ( dbFp ) )
      break ;
    year = getInt ( dbFp ) ;
    (void) getInt ( dbFp ) ;
    attrKey = getAttr ( dbFp ) ;
    title = mapTitleKeyToText ( titleKey, titleKeyFp, titleIndexFp ) ;
    dbAttr = mapAttrKeyToText ( attrKey, attrKeyFp, attrIndexFp ) ;
    keepFlag = alistMoviesOnlyTest ( mvsonly, title ) ;
    keepFlag = keepFlag && alistAttrTest ( attrFlags, attr, substring, casesen, dbAttr ) ;
    if ( yr )
    {
      if ( year < yrf || year > yrt )
        keepFlag = FALSE ;
      if ( keepFlag )
      {
        results [ count ] . title = title ;
        results [ count ] . attr = dbAttr ;
        results [ count ] . year = year ;
        count++ ;
      }
    }
    else 
    {
      if ( keepFlag )
        printf ( "%s %s\n", title, dbAttr ) ;
      free ( (void*) title ) ;
      free ( (void*) dbAttr ) ;
    }
  }

  if ( yr && count > 0 )
  {
    displayAlistTitleSorted ( results, count ) ;
    for ( i = 0 ; i < count ; i++ )
    {
      free ( (void*) results [ i ] . title ) ;
      free ( (void*) results [ i ] . attr ) ;
    }
  }
  (void) fclose ( dbFp ) ;
  (void) fclose ( titleKeyFp ) ;
  (void) fclose ( titleIndexFp ) ;
  (void) fclose ( attrKeyFp ) ;
  (void) fclose ( attrIndexFp ) ;
}


int main ( int argc, char **argv )
{
  int   err = FALSE ;
  int   i, j ;
  int   okopt ;
  int   casesen = FALSE ;
  int   substring = FALSE ;
  int   mvsonly = FALSE ;
  struct titleDbRec *years  ;
  char  attr [ MXLINELEN ] ;
  struct attrSearchOptRec  attrFlags = { 0, 0, 0, 0, 0 } ;
  int   yrf = 0 ;
  int   yrt = 9999 ;
  int   yropt = FALSE ;
  TitleID nyears = 0 ;
  int   listId = -1 ;
  int   titleFlag = FALSE ;

  years = (struct titleDbRec *) calloc ( MAXTITLES, sizeof ( struct titleDbRec ) ) ;
  if ( years == NULL )
    moviedbError ( "mkdb: not enough memory to generate titleDbRec" ) ;

  logProgram ( argv [ 0 ] ) ;

  if ( argc == 1 )
    err = TRUE ;

  attr [ 0 ] = '\0' ;

  for ( i = 1; i < argc; i++ )
    if ( strcmp ( "-i", argv[i] ) == 0 )
      casesen = TRUE ;
    else if ( strcmp ( "-m", argv[i] ) == 0 )
      mvsonly = TRUE ;
    else if ( strcmp ( "-s", argv[i] ) == 0 )
      substring = TRUE ;
    else if ( strcmp ( "-title", argv[i] ) == 0 )
      titleFlag = TRUE ;
    else if ( strcmp ( "-attr", argv[i] ) == 0 )
    {
      attrFlags . string = TRUE ;
      if ( ++i < argc )
	(void) strcpy ( attr, argv[i] ) ;
      else
	err = TRUE ;
    }
    else if ( strcmp ( "-yr", argv[i] ) == 0 )
      yropt = TRUE ;
    else if ( strcmp ( "-aa", argv[i] ) == 0 )
      attrFlags . aa = TRUE ;
    else if ( strcmp ( "-aan", argv[i] ) == 0 )
      attrFlags . aan = TRUE ;
    else if ( strcmp ( "-gg", argv[i] ) == 0 )
      attrFlags . gg = TRUE ;
    else if ( strcmp ( "-ggn", argv[i] ) == 0 )
      attrFlags . ggn = TRUE ;
    else if ( strcmp ( "-yrto", argv[i] ) == 0 )
    {
      yropt = TRUE ;
      if ( ++i < argc )
        yrt = atoi ( argv[i] ) ;
      else
        err = TRUE ;
    }
    else if ( strcmp ( "-yrfrom", argv[i] ) == 0 )
    {
      yropt = TRUE ;
      if ( ++i < argc )
        yrf = atoi ( argv[i] ) ;
      else
        err = TRUE ;
    }
    else if ( strcmp ( "-yreq", argv[i] ) == 0 )
    {
      yropt = TRUE ;
      if ( ++i < argc )
      {
        yrf = atoi ( argv[i] ) ;
        yrt = yrf ;
      }
      else
        err = TRUE ;
    }
    else
    {
      okopt = FALSE ;
      for ( j = 0 ; j < NO_OF_FILMOGRAPHY_LISTS ; j++ )
	if ( strcmp ( filmographyDefs [ j ] . option, argv [ i ] ) == 0 )
        {
          listId = j ;
	  okopt = TRUE ;
        }
      if ( ! okopt )
      {
         (void) fprintf ( stderr, "alist: unrecognised option %s\n", argv[i] ) ;
         err = TRUE ;
      }
    }

  if ( !attrFlags . string && !attrFlags . aa && !attrFlags . aan && !attrFlags . gg && !attrFlags . ggn )
  {
    (void) fprintf ( stderr, "alist: missing attribute to search for\n" ) ;
    err = TRUE ;
  }
  
  if ( !err && titleFlag == FALSE && listId < -1 )
  {
    (void) fprintf ( stderr, "alist: must pick -title or one database option\n" ) ;
    err = TRUE ;
  }

  if ( !err && titleFlag && listId >= 0 )
  {
    (void) fprintf ( stderr, "alist: can't pick -title and a database option\n" ) ;
    err = TRUE ;
  }

  if ( err )
    moviedbUsage ( ALIST_USAGE1, ALIST_USAGE2, ALIST_USAGE3, NULL, NULL, NULL ) ;

  if ( yropt )
    nyears = readTitleDb ( years ) ;

  if ( listId >= 0 )
    alistNames ( listId, attr, substring, casesen, attrFlags, mvsonly, yropt, yrf, yrt, nyears, years ) ;
  else
    alistTitles ( attr, substring, casesen, attrFlags, mvsonly, yropt, yrf, yrt ) ;

  return ( 0 ) ;
}
